crates/ruff/src/noqa.rs

```diff
@@ -256,38 +257,148 @@ enum ParsedFileExemption<'a> {
+            let line = ParsedFileExemption::lex_whitespace(line);
+            let line = ParsedFileExemption::lex_noqa(line)?;
+
+            if line.is_empty() {
+                // Ex) `# ruff: noqa`
+                Some(Self::All)
+            } else {
+                // Ex) `# ruff: noqa: F401, F841`
+                let line = ParsedFileExemption::lex_whitespace(line);
+                let Some(line) = ParsedFileExemption::lex_char(line, ':') else {
+                    warn!("Unexpected suffix on `noqa` directive: \"{line}\"");
+                    return None;
+                };
+                let line = ParsedFileExemption::lex_whitespace(line);
+
+                // Extract the codes from the line (e.g., `F401, F841`).
+                let mut codes = vec![];
+                let mut line = line;
+                while let Some(code) = ParsedFileExemption::lex_code(line) {
+                    codes.push(code);
+                    line = &line[code.len()..];
+
+                    // Codes can be comma- or whitespace-delimited.
+                    if let Some(rest) = ParsedFileExemption::lex_delimiter(line)
+                        .map(ParsedFileExemption::lex_whitespace)
+                    {
+                        line = rest;
+                    } else {
+                        break;
+                    }
+                }
 
-        if line.starts_with("# flake8: noqa")
-            || line.starts_with("# flake8: NOQA")
-            || line.starts_with("# flake8: NoQA")
-        {
-            return Some(Self::All);
+                Some(Self::Codes(codes))
+            }
+        } else {
+            None
         }
+    }
+
+    /// Lex optional leading whitespace.
+    #[inline]
+    fn lex_whitespace(line: &str) -> &str {
+        line.trim_start()
+    }
+
+    /// Lex a specific character, or return `None` if the character is not the first character in
+    /// the line.
+    #[inline]
+    fn lex_char(line: &str, c: char) -> Option<&str> {
+        let mut chars = line.chars();
+        if chars.next() == Some(c) {
+            Some(chars.as_str())
+        } else {
+            None
+        }
+    }
+
+    /// Lex the "flake8" prefix of a `noqa` directive.
+    #[inline]
+    fn lex_flake8(line: &str) -> Option<&str> {
+        line.strip_prefix("flake8")
+    }
 
-        if let Some(remainder) = line
-            .strip_prefix("# ruff: noqa")
-            .or_else(|| line.strip_prefix("# ruff: NOQA"))
-            .or_else(|| line.strip_prefix("# ruff: NoQA"))
+    /// Lex the "ruff" prefix of a `noqa` directive.
+    #[inline]
+    fn lex_ruff(line: &str) -> Option<&str> {
+        line.strip_prefix("ruff")
+    }
+
+    /// Lex a `noqa` directive with case-insensitive matching.
+    #[inline]
+    fn lex_noqa(line: &str) -> Option<&str> {
+        let mut chars = line.chars();
+        if chars
+            .next()
+            .map_or(false, |c| c.to_ascii_lowercase() == 'n')
```

Comment:
I'm surprised there isn't a better way to write this, but at least i found the issue for the missing methos https://github.com/rust-lang/rfcs/issues/2566

Reply:
We could match on the byte slice because we only compare with byte characters. But you then still need to do the invariant which is annoying?

```
match line.as_str().bytes() {
	['n' | 'N', 'o' | 'O', 'q' | 'Q' | 'a' | 'A' => tada,
	_ => nope
}
```

Reply:
the problem is that we want to select the string after `"noqa"` and converting back from bytes isn't safe.

Which reminds me, would `line.strip_prefix("noqa")` work?

Reply:
I don't think `line.strip_prefix("noqa")` matches case-insensitively, right?

Reply:
> the problem is that we want to select the string after "noqa" and converting back from bytes isn't safe.

I think it would be safe here right because we know that all variations of `noqa` have an exact length of 4 bytes

Anyway. I think the current code is fine. it's verbose but does the job

Reply:
I like this, I changed it.